### ISSUE TITLE
JIT/Methodical/Boxing/misc/concurgc.il failing in CI with a timeout

### ISSUE THREAD

**Issue by adityamandaleeka** [dotnet/coreclr#10022]
Hit this on a PR I opened, and saw that it also happened on another recent job. Here are the two instances:

https://ci.dot.net/job/dotnet_coreclr/job/master/job/x86_checked_windows_nt_prtest/3373/
https://ci.dot.net/job/dotnet_coreclr/job/master/job/x86_checked_windows_nt_prtest/3369/

```
cmdLine:D:\j\workspace\x86_checked_w---b3a226f6\bin\tests\Windows_NT.x86.Checked\JIT\Methodical\Boxing\misc\_relconcurgc_il\_relconcurgc_il.cmd Timed Out
18:42:02         Test Harness Exitcode is : -100
```

**Comment by gkhanna79** [comment]
CC @RussKeldorph

**Comment by RussKeldorph** [comment]
I've seen this time out in several x86 jobs but I can't repro locally.  Assuming it is a test issue, but worth at least cursory investigation.

**Comment by pgavlin** [comment]
Looking into it now.

**Comment by pgavlin** [comment]
I can't repro this either, but I did notice that the test is not using volatile reads and writes where it should be. #10072 fixes those issues; once it's merged I plan to close this bug until it hits again.

**Comment by mikedn** [comment]
Another one: https://ci.dot.net/job/dotnet_coreclr/job/master/job/x86_checked_windows_nt_prtest/3883/

**Comment by pgavlin** [comment]
For posterity, the crash dump for that failed build is located here: https://dumpling.azurewebsites.net/api/dumplings/archived/1d2e8c8f93185b2c93a36929f499d2d307595470

Unfortunately the test job did not upload artifacts so we do not have symbols for the dump.

However, I have looked at another dump from a different instance of this failure, and here is what seems to be happening:
- Thread A instantiates a `Thread` object and calls `Thread.Start`, which kicks off thread B
- Thread B immediately invokes the GC using `GC.Collect`
- Thread A ends up [blocked on `GCHeap:: WaitForGCEvent` waiting for the GC to complete](https://github.com/dotnet/coreclr/blob/master/src/gc/gcee.cpp#L389-L428) with `Thread.Start` on the stack
- Thread B completes its GC and immediately begins another, and so on

@jkotas @Maoni0 @swgillespie: it almost seems like the second thread (which does nothing but call `GC.Collect` in a tight loop) is starving the first thread from ever finishing its wait, but that is incompatible with my understanding of Win32 event semantics (viz. I would expect that setting an event with waiting threads would release at least one thread). Any ideas?

FWIW, I do not see very much value in this test. I think that it is attempting to find GC holes, which we have established and well-understood methods for doing so (GCStress). If the starvation behavior described above is expected, it is probably worth either rate-limiting the calls to `GC.Collect` or removing them entirely and leaning on GCStress to verify the GC info emitted for the box/unbox operations.

**Comment by jkotas** [comment]
There is no tight loop. The thread B waits 100ms between the GCs: https://github.com/dotnet/coreclr/blob/master/tests/src/JIT/Methodical/Boxing/misc/concurgc.il#L29

100ms should be a plenty of time for thread A to make progress.

**Comment by pgavlin** [comment]
Ah, great--I overlooked that.

I'm happy to provide the crash dump I have along with binaries and symbols if that would help, but it will take me a bit of time to do so (need to drop into the office and turn my workstation back on). That dump is definitely in the state described above. It is also possible that the GC simply is not able to complete, but AFAICT the dump does not contain enough information to determine how long it has been running for.

**Comment by jkotas** [comment]
It can wait till Monday.

The test is carefully crafted to store the iteration counter in a boxed object on GC heap. The place to look at would be the state of the iteration counter.

One possible theory is that the iteration counter is getting corrupted because of GC hole and the main thread keeps looping infinitely.

**Comment by Maoni0** [comment]
if you could provide a dump with symbols that would be really helpful. thanks!

**Comment by pgavlin** [comment]
@jkotas: in the dump I have (which was taken after the test had been running for 10 minutes), the main thread has not yet returned from its call to `Thread.Start`, so it does not fit the scenario you've described.

**Comment by jkotas** [comment]
Hit again in https://github.com/dotnet/coreclr/pull/10488

**Comment by jkotas** [comment]
@pgavlin Could you please email me link to the dump with symbols if you still have one?

**Comment by pgavlin** [comment]
@jkotas done.

**Comment by jkotas** [comment]
The stackwalker seems to be getting stuck in infinite loop. Looks like a recent regression, likely from Linux x86 stackwalking changes.

**Comment by jkotas** [comment]
The problem is that Frame linked list has cycle. m_Next is pointing to the frame itself.

It is caused by bad codegen for the `Func()` method in the test. The problem is that the part of the inlined PInvoke prolog gets included in the loop. When the loop executes second time, we get the Frame* cycle as the PInvoke prolog gets executed second time:

```
00db0908 55              push    ebp
00db0909 8bec            mov     ebp,esp
00db090b 57              push    edi
00db090c 56              push    esi
00db090d 53              push    ebx
00db090e 83ec1c          sub     esp,1Ch

00db0911 8d7ddc          lea     edi,[ebp-24h] // This is Frame*: ebp - 24: this, 
                                                                  ebp - 20: next
                                                                  ebp - 1c: datum
                                                                  ebp - 18: callsiteSP
                                                                  ebp - 14: return Address
                                                                  ebp - 10: caller FP

00db0914 e8572dd1ff      call    00ac3670 // This will link the Frame into the linked list
00db0919 8bc5            mov     eax,ebp
00db091b 8945f0          mov     dword ptr [ebp-10h],eax
00db091e 6a02            push    2
00db0920 6aff            push    0FFFFFFFFh
00db0922 c745e498c5ee6f  mov     dword ptr [ebp-1Ch],offset System_Private_CoreLib_ni+0x18c598 (6feec598)
00db0929 8bc4            mov     eax,esp // Return SP
00db092b 8945e8          mov     dword ptr [ebp-18h],eax
00db092e 8d054109db00    lea     eax,ds:[0DB0941h] // Return PC
00db0934 8945ec          mov     dword ptr [ebp-14h],eax
00db0937 c6460800        mov     byte ptr [esi+8],0
00db093b ff15604bfa6f    call    dword ptr [System_Private_CoreLib_ni+0x244b60 (6ffa4b60)] // GC collect
00db0941 c6460801        mov     byte ptr [esi+8],1
00db0945 833dac9f687100  cmp     dword ptr [coreclr!g_TrapReturningThreads (71689fac)],0
00db094c 7405            je      00db0953
00db094e e84de3ff6f      call    coreclr!JIT_RareDisableHelper (70daeca0)
00db0953 33c9            xor     ecx,ecx
00db0955 894dec          mov     dword ptr [ebp-14h],ecx
00db0958 b964000000      mov     ecx,64h
00db095d e8fea38a6f      call    System_Private_CoreLib_ni+0x8fad60 (7065ad60) // Sleep
00db0962 803d224dad0000  cmp     byte ptr ds:[0AD4D22h],0
00db0969 74a6            je      00db0911 // BUG BUG: This should be jumping to 00db091e instead

00db096b c6460801        mov     byte ptr [esi+8],1
00db096f 8b45e0          mov     eax,dword ptr [ebp-20h] // This will unlink the Frame from the linked list
00db0972 89460c          mov     dword ptr [esi+0Ch],eax
00db0975 8d65f4          lea     esp,[ebp-0Ch]
00db0978 5b              pop     ebx
00db0979 5e              pop     esi
00db097a 5f              pop     edi
00db097b 5d              pop     ebp
00db097c c3              ret
```

**Comment by pgavlin** [comment]
Thanks for investigating this @jkotas. The fix should be pretty straightforward: instead of always inserting the PInvoke prolog into the first basic block, we should create a new preceding BB iff the first BB has predecessors beyond the method prolog.

@BruceForstall @briansull @dotnet/jit-contrib: do you know of any other code that we unconditionally insert into the first BB that is logically part of the prolog?

**Comment by BruceForstall** [comment]
We have `fgEnsureFirstBBisScratch()` to do stuff like this.

**Comment by pgavlin** [comment]
Perfect. Thanks Bruce!

**Comment by pgavlin** [comment]
CLosed by #10587.
